Kernel/Thread.cpp

```diff
@@ -710,56 +708,107 @@ DispatchSignalResult Thread::dispatch_signal(u8 signal)
+    // We now place the thread state on the userspace stack.
+    // Note that we use a RegisterState.
+    // Conversely, when the thread isn't blocking the RegisterState may not be
+    // valid (fork, exec etc) but the tss will, so we use that instead.
+    auto& regs = get_register_dump_from_stack();
+
+    setup_signal_stack(regs, signal, old_signal_mask, handler_vaddr);
+    regs.eip = g_return_to_ring3_from_signal_trampoline.get();
 
 #ifdef SIGNAL_DEBUG
-        klog() << "signal: setting up user stack to return to eip: " << String::format("%p", ret_eip) << " esp: " << String::format("%p", old_esp);
+    klog() << "signal: Okay, " << *this << " {" << state_string() << "} has been primed with signal handler " << String::format("%w", m_tss.cs) << ":" << String::format("%x", m_tss.eip) << " to deliver " << signal;
 #endif
+    return DispatchSignalResult::Continue;
+}
 
-        // Align the stack to 16 bytes.
-        // Note that we push 56 bytes (4 * 14) on to the stack,
-        // so we need to account for this here.
-        u32 stack_alignment = (*stack - 56) % 16;
-        *stack -= stack_alignment;
+void Thread::setup_signal_stack(RegisterState& state, u8 signal, u32 old_signal_mask, VirtualAddress handler_vaddr)
+{
+    u32* stack = &state.userspace_esp;
+    u32 old_esp = *stack;
+    u32 ret_eip = state.eip;
+    u32 ret_eflags = state.eflags;
 
-        push_value_on_user_stack(stack, ret_eflags);
+#ifdef SIGNAL_DEBUG
+    klog() << "signal: setting up user stack to return to eip: " << String::format("%p", ret_eip) << " esp: " << String::format("%p", old_esp);
+#endif
 
-        push_value_on_user_stack(stack, ret_eip);
-        push_value_on_user_stack(stack, state.eax);
-        push_value_on_user_stack(stack, state.ecx);
-        push_value_on_user_stack(stack, state.edx);
-        push_value_on_user_stack(stack, state.ebx);
-        push_value_on_user_stack(stack, old_esp);
-        push_value_on_user_stack(stack, state.ebp);
-        push_value_on_user_stack(stack, state.esi);
-        push_value_on_user_stack(stack, state.edi);
+    // FIXME: Is there a way to avoid putting the extra 60 bytes for SIGINFO
+    //        onto every single signal handler stack, even when siginfo was not specified?
 
-        // PUSH old_signal_mask
-        push_value_on_user_stack(stack, old_signal_mask);
+    // FIXME: Plumb sender information into this part. We need to know more about the signal.
+    //        Maybe need a struct SignalReason?
+    auto signal_info = siginfo_for_signal(signal, nullptr);
 
-        push_value_on_user_stack(stack, signal);
-        push_value_on_user_stack(stack, handler_vaddr.get());
-        push_value_on_user_stack(stack, 0); //push fake return address
+    ucontext_t user_context {};
+    user_context.uc_link = nullptr; // TODO: Are we expected to maintain a linked list of these things in the kernel? in userspace?
+    user_context.uc_mcontext = 0;   // TODO: What is this for anyway? Dr Posix pls
```

Comment:
IIRC this is for the register dump.

Reply:
Gotcha yeah, figured that out last night and made an edit to the PR description :D

Probably the proper way to do this would be to have the register dump that is used in `sys$sigreturn` be the one from the ucontext?

Reply:
That's what I get for not reading the "Reminder to self" lol